### PR TITLE
Feature/add feature flag module

### PR DESCRIPTION
**Pull request by sechegaray**
Add feature flag module to allow enabling cdap-wide features.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml [hunk 1]
```diff
@@ -36,6 +36,11 @@
       <artifactId>cdap-api</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+      <version>${guava.version}</version>
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/pom.xml]: Why is this needed?

File: cdap-app-fabric/pom.xml [hunk 2]
```diff
@@ -253,6 +253,17 @@
...
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-features</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api</artifactId>
+      <version>${project.version}</version>
+      <scope>provided</scope>
```
Comment (tivv) [inline cdap-app-fabric/pom.xml]: Why is this needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java [hunk 3]
```diff
@@ -169,6 +171,34 @@
   protected final DynamicDatasetCache datasetCache;
   protected final RetryStrategy retryStrategy;
 
+
+  /**
+   * Returns runtime arguments as a Map by:
+   *  If INCLUDE_FEATURE_FLAGS_IN_RUNTIME_ARGUMEMTS: extracting any feature flag defined in CConfiguration
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java]: Use proper `<ol>` to ensure it's rendered properly

File: cdap-app-templates/cdap-etl/cdap-etl-core/pom.xml [hunk 4]
```diff
@@ -55,6 +55,12 @@
       <artifactId>cdap-proto</artifactId>
       <version>${project.version}</version>
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-core/pom.xml]: Why is it needed? I don't see test changes

File: cdap-app-templates/cdap-etl/hydrator-spark-core2_2.11/pom.xml [hunk 5]
```diff
@@ -87,6 +87,12 @@
       <artifactId>spark-sql_2.11</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/hydrator-spark-core2_2.11/pom.xml]: Why is it needed? I don't see test changes

File: cdap-app-templates/cdap-etl/hydrator-spark-core3_2.12/pom.xml [hunk 6]
```diff
@@ -87,6 +87,12 @@
       <artifactId>spark-sql_2.12</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/hydrator-spark-core3_2.12/pom.xml]: Why is it needed? I don't see test changes

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java [hunk 7]
```diff
@@ -191,8 +221,7 @@ protected AbstractContext(Program program, ProgramOptions programOptions, CConfi
     this.remoteClientFactory = remoteClientFactory;
     this.programRunId = program.getId().run(ProgramRunners.getRunId(programOptions));
     this.triggeringScheduleInfo = getTriggeringScheduleInfo(programOptions);
-
-    Map<String, String> runtimeArgs = new HashMap<>(programOptions.getUserArguments().asMap());
+    Map<String, String> runtimeArgs = getRuntimeArgs(programOptions);
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java:218]: This provides access to feature flags during the run from io.cdap.cdap.api.RuntimeContext#getRuntimeArguments. But we must ensure it's available on configuration / deploy time as well, e.g. in `io.cdap.cdap.etl.api.PipelineConfigurable#configurePipeline` and probably from `io.cdap.cdap.etl.api.StageContext`

File: cdap-features/src/main/java/io/cdap/cdap/common/features/Feature.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  Feature(String versionDeployed, boolean defaultAfterDeployment) {
+    this.featureFlagString = FEATURE_FLAG_PREFIX + this.name().toLowerCase().replace('_', '.') + ".enabled";
+    this.versionIntroduced = new PlatformInfo.Version(versionDeployed);
+    this.defaultAfterDeployment = defaultAfterDeployment;
+  }
+
+  public boolean isEnabled(StageContext context) {
+    Arguments args = context.getArguments();
```
Comment (tivv) [inline cdap-features/src/main/java/io/cdap/cdap/common/features/Feature.java]: This reads from `io.cdap.cdap.etl.common.AbstractStageContext#arguments`, not from `io.cdap.cdap.internal.app.runtime.AbstractContext#getRuntimeArguments`, is not it?

**Review by tivv — COMMENTED** [review]
Please add tests to ensure the data is propagated as expected

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/RemoteValidationTask.java [hunk 9]
```diff
@@ -105,7 +107,14 @@ public void run(RunnableTaskContext context) throws Exception {
       macroProperties -> systemAppContext
         .evaluateMacros(namespace, macroProperties, macroEvaluator, macroParserOptions);
     PluginConfigurer pluginConfigurer = systemAppContext.createPluginConfigurer(namespace);
-    StageValidationResponse validationResponse = ValidationUtils.validate(validationRequest, pluginConfigurer, macroFn);
+    StageValidationResponse validationResponse = ValidationUtils.validate(
+      validationRequest, pluginConfigurer, macroFn,
+      new FeatureFlagsProvider() {
+        public Map<String, String> getFeatureFlags() {
+          //Feature.extractFeatureFlags(getContext().getRuntimeArguments());
+          return Collections.emptyMap();
```
Comment (sechegaray) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/RemoteValidationTask.java]: I need to change this to get featureFlags

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java [hunk 10]
```diff
@@ -55,20 +56,22 @@
...
-   * @param macroFn           {@link Function} for evaluating macros
-   * @return {@link StageValidationResponse} in json format
-   */
+    /**
+     * Validate plugin based on the {@link StageValidationRequest}
+     *
+     * @param validationRequest {@link StageValidationRequest} with plugin properties
+     * @param pluginConfigurer  {@link PluginConfigurer} for using the plugin
+     * @param macroFn           {@link Function} for evaluating macros
+     * @return {@link StageValidationResponse} in json format
```
Comment (sechegaray) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationUtils.java]: update documentation

File: cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationConfigurer.java [hunk 11]
```diff
@@ -119,4 +122,6 @@ default void emitMetadata(Metadata metadata, MetadataScope scope) {
    * @return The {@link TriggerFactory} used to get triggers
    */
   TriggerFactory getTriggerFactory();
+
+  Map<String, String> getFeatureFlags();
```
Comment (tivv) [inline cdap-api/src/main/java/io/cdap/cdap/api/app/ApplicationConfigurer.java]: This is not needed given we have `FeatureFlagsProvider`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/DefaultAppConfigurer.java [hunk 12]
```diff
@@ -171,6 +180,8 @@ public void addSpark(Spark spark) {
     sparks.put(spec.getName(), spec);
   }
 
+
+
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/DefaultAppConfigurer.java]: Not needed

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java [hunk 13]
```diff
@@ -174,7 +176,12 @@ private void validateLocally(HttpServiceRequest request, HttpServiceResponder re
     Function<Map<String, String>, Map<String, String>> macroFn =
       macroProperties -> getContext().evaluateMacros(namespace, macroProperties, macroEvaluator, macroParserOptions);
     String validationResponse = GSON.toJson(ValidationUtils.validate(
-      validationRequest, getContext().createServicePluginConfigurer(namespace), macroFn));
+      validationRequest, getContext().createServicePluginConfigurer(namespace), macroFn,
+      new FeatureFlagsProvider() {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/ValidationHandler.java]: Looks like a lambda

File: cdap-features/src/main/java/io/cdap/cdap/features/Feature.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+/**
+ * Defines Features Flags to be used in CDAP.
+ * Features take the version that they were introduced as a first parameter. Optionally they can take a
+ * second paramenter to define their default behavior if they are not present in configuration.
+ * By default features default to enabled after they are introduced, and disabled before they were introduced
+ */
+public enum Feature {
+  ALLOW_FEATURE_FLAGS("6.6.0");
+  
+  public static final String FEATURE_FLAG_PREFIX = "feature.";
```
Comment (tivv) [inline cdap-features/src/main/java/io/cdap/cdap/features/Feature.java]: I'd move this to `FeatureFlagsProvider` and `extractFeatureFlags` to some utility class. This way we won't need a dependency on features module by the platform. Basically, Feature enum will become a library only for feature flag user

File: cdap-api/src/main/java/io/cdap/cdap/api/FeatureFlagsProvider.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ */
+
+package io.cdap.cdap.api;
+
+import java.util.Map;
+
+public interface FeatureFlagsProvider {
+
+  Map<String, String> getFeatureFlags();
```
Comment (tivv) [inline cdap-api/src/main/java/io/cdap/cdap/api/FeatureFlagsProvider.java:30]: Add javadocs to the interface and the method

File: cdap-app-fabric/pom.xml [hunk 16]
```diff
@@ -253,6 +253,11 @@
       <artifactId>spark-core_2.11</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-features</artifactId>
```
Comment (tivv) [inline cdap-app-fabric/pom.xml:258]: Do you still need this?

File: cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/service/ValidationUtilsTest.java [hunk 17]
```diff
@@ -55,6 +56,15 @@
  */
 public class ValidationUtilsTest {
 
+  private FeatureFlagsProvider getFeatureFlagsProvider() {
+    return new FeatureFlagsProvider() {
+      public Map<String, String> getFeatureFlags() {
+        return Collections.emptyMap();
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/test/java/io/cdap/cdap/datapipeline/service/ValidationUtilsTest.java]: Replace with lambda

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java [hunk 18]
```diff
@@ -169,6 +172,27 @@
   protected final DynamicDatasetCache datasetCache;
   protected final RetryStrategy retryStrategy;
 
+
+  /**
+   * Returns runtime arguments as a Map.
+   * <ol>
+   *  <li>If INCLUDE_FEATURE_FLAGS_IN_RUNTIME_ARGUMENTS is set: extracting any feature flag defined in CConfiguration
```
Comment (tivv) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java]: I think the feature name is different. I like this one more as it controls only runtime argument thing

Comment (tivv) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java]: May be not al long, but something that includes `RUNTIME` would be good

File: cdap-app-fabric/pom.xml (hunk 16 above)
Comment (sechegaray) [reply cdap-app-fabric/pom.xml:258]: yes, as we are using the feature in AbstractContext

**Review by tivv — APPROVED** [review]

File: cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+ * the License.
+ *
+ */
+
+package io.cdap.cdap.api.common;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class FeatureFlagsUtils {
```
Comment (chtyim) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: Missing javadoc.

Comment (chtyim) [reply cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: Also, `Util` without `s`

Comment (chtyim) [reply cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: For utility class, add a private no-op constructor to prevent instantiation.

File: cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+  public static final String FEATURE_FLAG_PREFIX = "feature.";
+
+  public static Map<String, String> extractFeatureFlags(Map<String, String> conf) {
```
Comment (chtyim) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: Missing javadoc for public method.

File: cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    Map<String, String> featureFlags = new HashMap<>();
+    for (String name : conf.keySet()) {
+      if (name.startsWith(FEATURE_FLAG_PREFIX)) {
+        String value = conf.get(name);
+        if (!(("true".equals(value) || ("false".equals(value))))) {
```
Comment (chtyim) [inline cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: Use `Boolean.parseBoolean` on the value instead of comparing string.

File: cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java (hunk 20 above)
Comment (chtyim) [reply cdap-api-common/src/main/java/io/cdap/cdap/api/common/FeatureFlagsUtils.java]: Is feature flag boolean? If so, the map returned should be `Map<String, Boolean>`? Also, it might be better to use a class instead of a Map so that new function can be added later easier.

File: cdap-app-fabric/pom.xml (hunk 16 above)
Comment (chtyim) [reply cdap-app-fabric/pom.xml:258]: What are we using in AbstractContext that belongs to the cdap-features module?

File: cdap-features/src/main/java/io/cdap/cdap/features/Feature.java [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  private boolean getDefaultValue() {
+    if (PlatformInfo.getVersion().compareTo(versionIntroduced) <= 0) {
+      return false;
+    } else {
+      return defaultAfterDeployment;
+    }
+  }
+
```
Comment (chtyim) [inline cdap-features/src/main/java/io/cdap/cdap/features/Feature.java:98]: Remove extra new lines

File: cdap-features/src/main/java/io/cdap/cdap/features/Feature.java [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+    this(versionIntroduced, true);
+  }
+
+  Feature(String versionDeployed, boolean defaultAfterDeployment) {
+    this.featureFlagString = FEATURE_FLAG_PREFIX + this.name().toLowerCase().replace('_', '.') + ".enabled";
+    this.versionIntroduced = new PlatformInfo.Version(versionDeployed);
+    this.defaultAfterDeployment = defaultAfterDeployment;
+  }
+
+  public boolean isEnabled(FeatureFlagsProvider featureFlagsProvider) {
```
Comment (chtyim) [inline cdap-features/src/main/java/io/cdap/cdap/features/Feature.java:58]: Missing javadoc for public methods.

File: cdap-features/src/main/java/io/cdap/cdap/features/Feature.java [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+  public static final String FEATURE_FLAG_PREFIX = FeatureFlagsUtils.FEATURE_FLAG_PREFIX;
+  private final PlatformInfo.Version versionIntroduced;
+  private final boolean defaultAfterDeployment;
+  private final String featureFlagString;
+
+  Feature(String versionIntroduced) {
+    this(versionIntroduced, true);
+  }
+
+  Feature(String versionDeployed, boolean defaultAfterDeployment) {
```
Comment (chtyim) [inline cdap-features/src/main/java/io/cdap/cdap/features/Feature.java]: What do mean by defaultAfterDeployment? What deployment?

File: cdap-app-fabric/pom.xml (hunk 16 above)
Comment (sechegaray) [reply cdap-app-fabric/pom.xml:258]: We are using a Feature Flag to guard adding the "feature." configuration options into runtime arguments.

File: cdap-api/src/main/java/io/cdap/cdap/api/workflow/WorkflowConfigurer.java [hunk 25]
```diff
@@ -28,7 +30,8 @@
 /**
  * Configurer for configuring the {@link Workflow}.
  */
-public interface WorkflowConfigurer extends ProgramConfigurer, PluginConfigurer, DatasetConfigurer {
+public interface WorkflowConfigurer
+  extends ProgramConfigurer, PluginConfigurer, DatasetConfigurer, FeatureFlagsProvider {
```
Comment (chtyim) [inline cdap-api/src/main/java/io/cdap/cdap/api/workflow/WorkflowConfigurer.java:34]: I think the `FeatureFlagsProvider` should extended by all program configurer interfaces, not just `Workflow`.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/DefaultAppConfigurer.java [hunk 26]
```diff
@@ -112,6 +120,7 @@ public DefaultAppConfigurer(Id.Namespace namespace, Id.Artifact artifactId, Appl
     this.pluginInstantiator = pluginInstantiator;
     this.appMetadata = new HashMap<>();
     this.triggerFactory = new DefaultTriggerFactory(namespace.toEntityId());
+    this.featureFlags = featureFlags;
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/DefaultAppConfigurer.java:123]: Do a defensive copy and make it immutable.

`Collections.unmodifiableMap(new HashMap<>(featureFlags));`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java [hunk 27]
```diff
@@ -169,6 +172,27 @@
...
+   * </ol>
+   *
+   * @param programOptions
+   * @return a map of runtime arguments.
+   */
+  private Map<String, String> getRuntimeArgs(ProgramOptions programOptions) {
+    Map<String, String> runtimeArgs = new HashMap<>();
+    Map<String, String> conf = CConfigurationUtil.asMap(cConf);
+    if (Feature.FEATURE_FLAGS_IN_RUNTIME.isEnabled(conf)) {
+      runtimeArgs.putAll(FeatureFlagsUtil.extractFeatureFlags(conf));
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/AbstractContext.java:190]: I don't think we should blend feature configs into the runtime arguments. We should have all program context extends from FeatureFlagProvider instead. A feature flags can be overridden by an argument from the runtime arguments.
